src/library/scala/collection/mutable/HashSet.scala

```diff
@@ -203,6 +202,41 @@ final class HashSet[A](initialCapacity: Int, loadFactor: Double)
     }
   }
 
+  override def filterInPlace(p: A => Boolean): HashSet.this.type = {
+    if (isEmpty) {
+      this
+    } else {
+      var bucket = 0
+
+      while (bucket < table.length) {
+        table(bucket) = table(bucket) match {
+          case null => null
+          case head =>
+            var newHead: Node[A] = null
+            var prev: Node[A] = null
+            var next = head
+            while (next ne null) {
+              if (p(next.key)) {
+                if (newHead eq null) {
```

Comment:
I think it would be faster yet if you split the loop in two; make `head` a var, and first do `while ((head ne null) && !p(head.key)) head = head.next`.  Then you can do the second part of the loop without having to check for `newHead` and stuff.